### ISSUE TITLE
perf(cuda): the multi-row speculative verify forward is launch-bound on GB10

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1782]
## Problem / Background

Measured under #1351 (PR #1771, still open) on GB10 sm_121: the Laguna XS 2.1 NVFP4 target with its 0.9 GB bf16 DFlash drafter, greedy, raw code prompts, 200 tokens, idle host, GPU held exclusively for the whole sweep, n=3 per configuration against an n=5 kill-switch-off baseline on the same binary. No speculative configuration wins.

| Configuration | n | tok/s mean (min to max) | Mean accepted | vs off |
|---|---|---|---|---|
| off (baseline) | 5 | 32.50 (31.24 to 33.58) | | |
| block 2 | 3 | 16.57 (16.38 to 16.87) | 0.84 | 0.51x |
| block 4 | 3 | 27.23 (27.01 to 27.52) | 2.21 | 0.84x |
| block 6 | 3 | 30.90 (30.55 to 31.38) | 2.90 | 0.95x |
| block 8 | 3 | 31.47 (31.10 to 31.98) | 3.33 | 0.97x |
| block 10 | 3 | 31.03 (30.86 to 31.18) | 3.42 | 0.95x |
| block 16 (checkpoint default) | 3 | 27.92 (27.36 to 28.45) | 3.55 | 0.86x |

Within-configuration spread is about 2 percent, so this is not the GB10 single-run decode bimodality from #755 hiding a win.

Acceptance is not the binding constraint, the verify block's device cost is. PR #1771 reports, per round at block 8, about 32 ms of host-side drafter graph construction, 3 ms of target graph construction and 100 ms of synchronized device work for 4.33 emitted tokens, against 30.8 ms per token on the classic path. Across widths the verify block's device cost fits a fixed **77 ms plus 3.3 ms per row** (83 ms at 2 rows, 130 ms at 16), so even a 2-row verify costs 2.7x a single-token step. With a fixed term that size, perfect acceptance at block 8 caps the reachable speedup near 1.8x, and the measured 3.3 to 3.6 accepted tokens land at 0.86x to 0.97x.

A per-block cost nearly independent of row count is a property of how the CUDA backend executes a multi-row forward, not of any one drafter. Every speculative family in the tree makes the same call, so fixing it once is the lever that makes speculative decoding worth enabling on this hardware.

## Current Behavior

One call per round, shared by every DFlash family: `src/lib/mlxcel-core/src/drafter/dflash/round_loop.rs:780` invokes `SpeculativeTarget::verify_forward_with_capture_layers` (trait at `:162` and `:177`) on a `[1, block_size]` input, then async-evals logits and captured hiddens at `:810`. The MTP loop has the same structure through `MtpTarget::verify_forward` (`src/lib/mlxcel-core/src/speculative/mtp/target.rs:303`). DFlash targets today are Qwen 3.5 and Qwen 3.5 VL, LFM2 and LFM2 VL, and Muse Glimmer text and VLM, registered for serving at `src/server/batch/dflash_target.rs:236` through `:341`; MTP adapters are Gemma 4 (text, VL, Unified, plus batched variants), Qwen 3.5, Inkling and GLM-4 MoE Lite, in `src/models/*_mtp_target.rs`. Laguna joins the DFlash list when #1771 merges. Only Laguna has been measured; the rest share the code path but not yet a number.

What is **not** established is that the multi-row forward goes uncaptured. In the pinned MLX tree (`src/lib/mlx-cpp/CMakeLists.txt:122`, `GIT_TAG 81ba1c6a`) nothing in the capture path is conditioned on row count: `CommandEncoder::CaptureContext` begins a stream capture for every eval unless `MLX_USE_CUDA_GRAPHS=0`, and `commit()` reuses a cached `cudaGraphExec` through `cudaGraphExecUpdate` under a structural key from `subgraph_to_key`, all in `mlx/backend/cuda/device.cpp`. Under #1545 the Volta decode path showed `cudaGraphInstantiate` saturating at 196 while `cudaGraphExecUpdate` kept growing with tokens, which is what replay looks like. So the fixed 77 ms must be attributed before it is fixed. Three candidates the pinned MLX makes concrete:

- `get_graph_limits` (`mlx/backend/cuda/device.cpp:181`) budgets a graph at 20 ops and **25 MB** for compute capability 12.1 (`case 1210: // DGX Spark`, this host), against 400 MB on A100 and 1000 MB on H100, B200 and consumer Blackwell, with 100 MB as the fallback. `set_input_array` adds every input's `data_size()` to `bytes_in_graph_` and `needs_commit()` trips on that budget, so a forward reading more bytes per node commits more, smaller graphs. GB10 has the tightest byte budget of any architecture MLX lists.
- `subgraph_to_key` clears `is_updatable` for any node type outside its switch and for a kernel whose cluster dim y or z exceeds 1. One such node anywhere in the multi-row path forces the re-instantiate branch of `commit()` instead of the update branch.
- A varying node key is a separate cache entry. Block width is fixed per run here, context length is not.

Prior art to read rather than re-derive: `apply_cuda_graph_cache_default` (`src/lib/mlxcel-core/src/hardware.rs:289`) raises `MLX_CUDA_GRAPH_CACHE_SIZE` to 2000 on CUDA builds (#818); `maybe_disable_cuda_graphs_for_model` (`src/loading/mod.rs:562`) sets `MLX_USE_CUDA_GRAPHS=0` for Gemma 4 only (#688), so Gemma 4 on CUDA has no capture to improve; `apply_metal_ops_per_buffer_default` (`src/lib/mlxcel-core/src/hardware.rs:222`) is the precedent for a hardware-gated buffer-limit default, applied on Metal only and leaving CUDA on MLX's values. `MLX_MAX_MB_PER_BUFFER` appears nowhere in this tree and is undocumented.

## Proposed Solution

**Phase 0, attribute the fixed 77 ms before changing anything.** Profile with `nsys profile -t cuda,nvtx --cuda-graph-trace=node` (mandatory, rule 4 of `docs/benchmark_results/volta-sm70-baseline-2026-08-31.md`) and report, for a single-token decode step and for verify blocks at 2 and 16 rows, the `cudaGraphInstantiate`, `cudaGraphExecUpdate`, `cudaGraphLaunch` and kernel counts. The question: does the multi-row commit re-instantiate where the single-token commit updates, and if so, which node denies updatability. Two zero-code controls belong here since either could settle it without a patch: `MLX_MAX_MB_PER_BUFFER` raised above GB10's 25 MB, and `MLX_MAX_OPS_PER_BUFFER` above 20, each against the same baseline. Report the answer even if it is that graph replay is not where the time goes; a measured negative closes this honestly and redirects the drafter work.

**Phase 1, make the verify block's device cost scale with rows.** The fix follows from phase 0: remove or replace whatever node denies updatability, key the captured graph per row count, or pad the block to one fixed width so a single captured graph serves every round. State in the PR which was chosen and why. The tradeoff is real: per-row-count keying multiplies distinct graph keys against the `MLX_CUDA_GRAPH_CACHE_SIZE` budget and the lifetime-miss abort of #818, while fixed-width padding pays for rows the round does not need and changes the arithmetic each row sees unless the padded rows are masked out exactly.

## Scope

**In scope:** attribution of the fixed per-block device cost, the fix that follows, any hardware-gated default or knob it needs (alongside `apply_cuda_graph_cache_default`, plus its row in `docs/environment-variables.md` and the help block at `src/main.rs:59`), and the re-measured block-size sweep. An MLX-side fix belongs under `src/lib/mlx-cpp/patches/mlx/backend/cuda/`, not as an edit to the fetched checkout.

**Out of scope:** the drafter's own 32 ms per round of host graph construction, the same class of cost #1713 measures on Metal and #1185 measured for MTP; acceptance-rate work on any drafter; Metal, where these numbers say nothing; and Gemma 4 on CUDA, unreachable by this fix until #688 is revisited.

## Implementation Notes

- **Reuse the counters that exist.** `DFlashDiagnostics` (`round_loop.rs:509`) already splits a round into `draft_time_ms`, `verify_time_ms`, `target_argmax_time_ms`, `walk_time_ms`, `hidden_concat_time_ms` and `rollback_time_ms`, and the server logs all of them on the `DFlash diagnostics` line at `src/server/batch/dflash_target.rs:639`. Mind the documented subtlety at `round_loop.rs:531`: MLX is lazy, so `verify_time_ms` is host graph construction and the device work lands in `target_argmax_time_ms`. Use `MLXCEL_PROFILE_PIPELINE_DETAIL` (`docs/environment-variables.md:504`) for the classic-path arm. Do not add parallel timers.
- **Constraint, where the sweep can run.** `mlxcel generate --draft-kind dflash` does not exist on `main` at `d51ab751`: `src/commands/generate.rs:1802` errors naming the missing round loop, and PR #1771 adds the offline arm for Laguna only. The Qwen 3.5 pairing is reachable today through `mlxcel-server`, and both halves are on this host (`models/mlx/qwen3.5-4b-4bit` with `models/mlx/qwen3.5-4b-dflash`, a real DFlash drafter carrying `block_size: 16` and `target_layer_ids: [1, 8, 15, 22, 29]`). Qwen 3.5 keeps the permissive `exactness_allows` default (`dflash_target.rs:197`) so the burst engages unaided, unlike Laguna, whose probe declines this host and needs `MLXCEL_MTP_ALLOW_INEXACT=1`.
- **Constraint, output must not move.** Any padding or masking scheme has to leave greedy ids byte-identical to the pre-change run on the same binary and prompt. A speed win paid for in different tokens is a regression.
- **Edge cases:** a partial accept leaves a shorter next verify; width 2 and the checkpoint's 16 must both work; a context crossing a KV-cache growth step changes shapes mid-run; and the batched burst (`drafter/dflash/round_loop_batched.rs`) varies rows by batch as well as by width. State what each does under the chosen scheme.
- **Error handling:** when a captured graph cannot be reused for a row count, fall back to the current path and log once at debug, never per round and never as a request failure. Watch #818's lifetime-miss abort, which kills the process, if the change multiplies distinct graph keys.

## Acceptance Criteria

- [x] The fixed per-block term is attributed to a named cause with profiler evidence, written up as a dated GB10 record under `docs/benchmark_results/`. Required even if the outcome is that graph replay is not the cause.
- [x] The verify block's device cost is reported at 2, 4, 8 and 16 rows so the fixed and per-row terms are separable, and the 2-row cost is within a stated factor of a single-token decode step on the same model and context, replacing the current 2.7x.
- [x] The #1351 sweep is re-run by the same method as the table above (same binary both arms, kill-switch A/B, idle host, GPU exclusive, n >= 3 per configuration, min to max per cell) and posted before and after in the same table shape.
- [x] At least one configuration beats the kill-switch-off baseline with its whole range outside the baseline's range, or the issue closes with a measured negative result stating the remaining floor.
- [x] A second family is measured on the same host, so the effect is shown to be a backend property rather than a Laguna one. The Qwen 3.5 server pairing above is the cheapest choice.
- [x] Greedy output is byte-identical before and after on every pairing touched, at the same block width.
- [x] The change lands on the real serving and offline decode path, not behind an opt-in flag nobody sets, and any new knob is documented in `docs/environment-variables.md` and listed in `src/main.rs`.
- [ ] `make verify-test-cuda` (`cargo test --workspace --profile test-fast --features cuda --no-fail-fast -- --test-threads=1`, per `CONTRIBUTING.md:62`), `cargo clippy --workspace --all-targets --features cuda -- -D warnings` and `cargo fmt --all -- --check` pass.

## Verification

```bash
cargo build --release --features cuda

# Baseline arm and speculative arm, same binary.
./target/release/mlxcel-server -m models/mlx/qwen3.5-4b-4bit --port 8080
./target/release/mlxcel-server -m models/mlx/qwen3.5-4b-4bit \
  --draft-model models/mlx/qwen3.5-4b-dflash --draft-kind dflash \
  --draft-block-size 8 --port 8080
# Per burst, read the `DFlash diagnostics` line: rounds, accepted_tokens,
# emitted_per_verify, draft_ms, verify_ms, target_argmax_sync_ms, decode_ms.

# Attribution. --cuda-graph-trace=node is mandatory or every kernel reads near zero.
nsys profile -t cuda,nvtx --cuda-graph-trace=node -o verify_b8 \
  ./target/release/mlxcel-server -m models/mlx/qwen3.5-4b-4bit \
  --draft-model models/mlx/qwen3.5-4b-dflash --draft-kind dflash --draft-block-size 8
nsys stats --report cuda_api_sum --format csv verify_b8.nsys-rep

# Zero-code controls against GB10's 20 ops / 25 MB graph budget.
MLX_MAX_MB_PER_BUFFER=400 ./target/release/mlxcel-server ...
MLX_MAX_OPS_PER_BUFFER=100 ./target/release/mlxcel-server ...

make verify-test-cuda
```

A pass looks like: `cudaGraphInstantiate` flat in round count on the verify path rather than growing, the 2-row verify close to a single-token step, at least one block width whose tok/s range sits entirely above the kill-switch-off range, and identical greedy ids on both arms. Every run states its PTX-cache and page-cache state, per rules 3 and 7 of `docs/benchmark_results/volta-sm70-baseline-2026-08-31.md`.

## Technical Considerations

Provenance: #1351 and PR #1771 carry the sweep and the per-round split. Related: #1545 measured graph instantiation, the graph cache and `MLX_MAX_OPS_PER_BUFFER` on Volta and found none of them where the time was, which is the strongest argument for running phase 0 before writing a fix; #1713 is the Metal-side neighbour where host graph build is serial with device execution; #1185 measured the MTP drafter's own per-step overhead; #818 bounds how many distinct graph keys are safe; #688 is why Gemma 4 on CUDA has no capture to improve. Host for every number above: GB10 sm_121, MLX pin `81ba1c6a`.
